subsys/settings/src/settings_eeprom.c

```diff
@@ -0,0 +1,483 @@
+			if (rc) {
+				break;
+			}
+			/* copy the id if required */
+			if ((id) && ((dataidend - *addr) > 0)) {
+				size_t cplen = MIN((dataidend - *addr), rdlen);
+
+				memcpy(id + (*addr - datastart), buf, cplen);
+			}
+			/* calculate the crc16 */
+			crc16cal = crc16_ccitt(crc16cal, buf, rdlen);
+			*addr += rdlen;
+		}
+
+		/* read the stored crc16 */
+		rc = eeprom_read(cf->eeprom, *addr, &crc16sto,
+				 sizeof(crc16sto));
+		if (rc) {
+			break;
+		}
+		*addr += sizeof(crc16sto);
+
+		if (crc16sto == crc16cal) {
+			/* update the info if needed */
+			if (info) {
+				info->idlen = dataidend - datastart;
+				info->dataoffset = dataidend;
+				info->datalen = dataend - dataidend;
+			}
+			/* set the return value to the record length */
+			rc = (int)reclen;
+			break;
+		}
+	}
+	return rc;
+}
+
+static int settings_eeprom_load(struct settings_store *cs,
+			     const struct settings_load_arg *arg)
+{
+	struct settings_eeprom *cf = (struct settings_eeprom *)cs;
+	struct settings_eeprom_read_fn_arg read_fn_arg;
+	struct settings_eeprom_rec_info info;
+	char id[SETTINGS_MAX_NAME_LEN + SETTINGS_EXTRA_LEN + 1];
+	off_t addr = cf->start + sizeof(struct settings_eeprom_hdr);
+	int rc = 0;
+
+	if (!device_is_ready(cf->eeprom)) {
+		return -EIO;
+	}
+
+	while (1) {
+
+		rc = settings_eeprom_recstep(cf, &addr, &info, (uint8_t *)id);
+		if (rc <= 0) {
+			break;
+		}
+
+		id[info.idlen] = '\0';
+
+		read_fn_arg.eeprom = cf->eeprom;
+		read_fn_arg.address = info.dataoffset;
+		read_fn_arg.len = info.datalen;
+
+		rc = settings_call_set_handler(id, info.datalen,
+					       settings_eeprom_read_fn,
+					       &read_fn_arg, (void *)arg);
+		if (rc) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+/* invalidate the record that ends at addr */
+static int settings_eeprom_invalidate(struct settings_eeprom *cf, off_t addr)
+{
+	uint16_t crc16inv;
+	int rc;
+
+	addr -= sizeof(crc16inv);
+	rc = eeprom_read(cf->eeprom, addr, &crc16inv, sizeof(crc16inv));
+	if (rc) {
+		return rc;
+	}
+	crc16inv = ~crc16inv;
+	return eeprom_write(cf->eeprom, addr, &crc16inv, sizeof(crc16inv));
+}
+
+/* Compress data in eeprom (remove deleted items) */
+static int settings_eeprom_compress(struct settings_eeprom *cf)
+{
+	off_t addr = cf->start + sizeof(struct settings_eeprom_hdr);
+	off_t wraddr = cf->start + sizeof(struct settings_eeprom_hdr);
+	uint16_t reclen;
+	uint8_t buf[4];
+	int rc = 0;
+
+	if (!device_is_ready(cf->eeprom)) {
+		return -EIO;
+	}
+
+	while (1) {
+		rc = settings_eeprom_recstep(cf, &addr, NULL, NULL);
+		if (rc <= 0) {
+			break;
+		}
+		reclen = (uint16_t)rc;
+		addr -= reclen;
+
+		while (reclen) {
+			size_t cplen = MIN(reclen, sizeof(buf));
+
+			rc = eeprom_read(cf->eeprom, addr, buf, cplen);
+			if (rc) {
+				break;
+			}
+			addr += cplen;
+
+			rc = eeprom_write(cf->eeprom, wraddr, buf, cplen);
+			if (rc) {
+				break;
+			}
+			wraddr += cplen;
+			reclen -= cplen;
+		}
+
+		/* invalidate the record at addr unless it is at wraddr */
```

Comment:
Maybe "invalidate already copied record unless it is record that has just been written"; it is not at the address, the address is actually pointing after the invalidated record. Comment is misleading.